File: matmul/matmul.rb
# Writen by Attractive Chaos; distributed under the MIT license
# frozen_string_literal: true

require 'socket'

def transposed(n_size, p_size, matrix_b)
  b2 = Array.new(n_size) { Array.new(p_size) { 0 } }
  n_size.times do |i|
    p_size.times do |j|
      b2[j][i] = matrix_b[i][j]
    end
  end
  b2
end

def multiplication(m_size, n_size, p_size, matrix_a, matrix_b2)
  c = Array.new(m_size) { Array.new(p_size) { 0 } }
  m_size.times do |i|
    p_size.times do |j|
      ai = matrix_a[i]
      b2j = matrix_b2[j]
      c[i][j] = n_size.times.reduce(0) { |s, k| s + ai[k] * b2j[k] }
    end
  end
  c
end

def matmul(matrix_a, matrix_b)
  m = matrix_a.length
  n = matrix_a[0].length
  p = matrix_b[0].length

  # transpose
  b2 = transposed(n, p, matrix_b)

  # multiplication
  multiplication(m, n, p, matrix_a, b2)
end

def matgen(num, seed)
  tmp = seed / num / num
  a = Array.new(num) { Array.new(num) { 0 } }
  num.times do |i|
    num.times do |j|
      a[i][j] = tmp * (i - j) * (i + j)
    end
  end
  a
end

def notify(msg)
  Socket.tcp('localhost', 9001) { |s| s.puts msg }
rescue SystemCallError
  # standalone usage
end

def calc(n)
  n = n >> 1 << 1
  a = matgen(n, 1.0)
  b = matgen(n, 2.0)
  c = matmul(a, b)
  c[n >> 1][n >> 1]
end

if __FILE__ == $PROGRAM_NAME
  n = ARGV.length.positive? ? ARGV[0].to_i : 100

  left = calc(101)
  right = -18.67
  if (left - right).abs > 0.1
    warn "#{left} != #{right}"
    exit(1)
  end

  engine = RUBY_ENGINE
  if engine == 'truffleruby'
    desc = RUBY_DESCRIPTION
    if desc.include?('Native')
      engine = 'Ruby/truffleruby'
    elsif desc.include?('JVM')
      engine = 'Ruby/truffleruby (--jvm)'
    end
  elsif engine == 'ruby' && RubyVM::MJIT.enabled?
    engine = 'Ruby (--jit)'
  end

  notify("#{engine}\t#{Process.pid}")
  results = calc(n)
  notify('stop')

  puts results
end

File: matmul/matmul.pl
# rurban; distributed under the MIT license
use v5.32;
use warnings;
no feature qw(indirect);
use feature qw(signatures);
no warnings qw(experimental);

use Socket;

sub matmul ($a, $b) {
    my $m = $#{$a} - 1;
    my $n = $#{$a->[0]} - 1;
    my $p = $#{$b->[0]} - 1;

    # transpose
    my @b2;
    for my $i (0..$n) {
        for my $j (0..$p) {
            $b2[$j]->[$i] = $b->[$i]->[$j];
        }
    }

    # multiplication
    my @c;
    for my $i (0..$m) {
        for my $j (0..$p) {
            my $s = 0;
            my ($ai, $b2j) = ($a->[$i], $b2[$j]);
            for my $k (0..$n) {
                $s += $ai->[$k] * $b2j->[$k];
            }
            $c[$i]->[$j] = $s;
        }
    }
    \@c
}

sub matgen ($n, $seed) {
    my $tmp = $seed / $n / $n;
    my @a;
    for my $i (0..$n) {
        for my $j (0..$n) {
            $a[$i]->[$j] = $tmp * ($i - $j) * ($i + $j);
        }
    }
    \@a
}

sub notify ($msg) {
    socket(my $socket, Socket::PF_INET, Socket::SOCK_STREAM, (getprotobyname('tcp'))[2]);
    if (connect($socket, Socket::pack_sockaddr_in(9001, Socket::inet_aton('localhost')))) {
        print $socket $msg;
    }
    close($socket);
}

sub calc ($n) {
    $n = int($n / 2) * 2;
    my $a = matgen($n, 1.0);
    my $b = matgen($n, 2.0);
    my $c = matmul($a, $b);
    $c->[$n/2]->[$n/2]
}

if ($0 eq __FILE__) {
    my $n = @ARGV ? shift : 100;

    my $left = calc(101);
    my $right = -18.67;
    if (abs($left - $right) > 0.1) {
        print STDERR "${left} != ${right}\n";
        exit(1);
    }

    notify("Perl\t" . $$);
    my $results = calc($n);
    notify("stop");

    print $results, "\n";
}
